Add memory leak check in CUDA tests

ssnl opened the pull request:
With `torch.cuda.memory_allocated` available, we can test if CUDA methods have memory leaks. This PR adds a wrapper around each CUDA test. It checks if the CUDA memory usage before and after stay constant.

Also move `TEST_CUDA`, `TEST_MULTIGPU`, `TEST_CUDNN`, `TEST_CUDNN_VERSION` to `common_cuda.py`.

ssnl committed:
Add memory leak check in CUDA tests

ssnl committed:
Tracking multi-GPU too

soumith commented:
this is kinda cool!

ezyang commented:
Excellent. Have you purposely introduced a memory leak and verified that this test catches it?

ssnl committed:
fix run_test.py not running __name__ == '__main__' content; add test for make_cuda_memory_checked_test

SsnL commented:
@ezyang I added a test for the new method in new commit.

cc @goldsborough for changes to `run_test.py`.

ssnl committed:
add a comment

goldsborough commented:
Cool, just make sure the tests actually run in the CI, even when green

ssnl committed:
skip if cuda

ssnl committed:
1. Change the wrapper to a method in common.py:TestCase
2. Refactor common constants/method that initialize CUDA context into common_cuda.py
3. Update some test files to use TEST_CUDA and TEST_MULTIGPU

apaszke commented:
Does it have any effect on the perf of the tests?

SsnL commented:
In local tests and CI test, this new mechanism detected `MaxUnpool3d` (`test_nn.TestNN.test_MaxUnpool3d_net_cuda`) and `MultiLabelMarginCriterion` leaking memory (`test_legacy_nn.TestNN.test_MultiLabelMarginCriterion_1d_cuda`).

SsnL commented:
@apaszke I don't have concrete numbers on full test suite. But for `test_cuda` on devgpu, the total time went from `20s` to `58s`. The overall test time increase on CUDA tests is likely less than 3 fold because `test_cuda` tests are small and thus the wrapper overhead area more obvious.

Still, I think there are good reasons to get this in. The above detected leaks are good examples.

Once I fix those leaks, I will run the full suite and report the total increased time here.

ssnl committed:
Fix MaxUnpool3d forward memory leak

ssnl committed:
Fix MultiLabelMarginCriterion forward memory leak

ezyang commented:
Not for this PR, but we can do something similar for CPU leaks too; all we need to do is distinguish between tensor allocations (unlikely to stay live across tests) and metadata allocations (we won't track leaks for those.)

Although, when @gchanan has his way all the manual refcounting in TH will go away and it should become a lot harder to make this kind of mistake.

SsnL commented:
@pytorchbot retest this please

ssnl committed:
Fix MultiMarginLoss backward memory leak

albanD commented on test/common.py:
```diff
@@ -413,6 +419,33 @@ def accept_output(update_type):
+    def run_with_cuda_memory_check(self, fn, *args, **kwargs):
+        num_devices = torch.cuda.device_count()
+        gc.collect()
+        torch.cuda.synchronize()
```
Why do you perform first collect then synchronize? Is the order important or the two can't interact?

SsnL replied:
The order shouldn't matter. As long as the refcnts go to zero it should be fine.

albanD commented on test/test_cuda.py:
```diff
@@ -1626,6 +1626,28 @@ def test_overlapped_indices(self):
+        with self.assertRaisesRegex(AssertionError, r"leaked \d+ bytes CUDA memory on device 0"):
+            self.run_with_cuda_memory_check(leak_gpu0)
+
+        if torch.cuda.device_count() > 1:
```
You can use "TEST_MULTIGPU" here no?

apaszke commented on aten/src/THCUNN/generic/MultiLabelMarginCriterion.cu:
```diff
@@ -87,6 +87,7 @@ void THNN_(MultiLabelMarginCriterion_updateOutput)(
 
   THCTensor_(free)(state, input);
   THCIndexTensor_(free)(state, target);
+  THCTensor_(free)(state, istarget);
```
Those modules must be used a lot haha. Maybe we should move their implementation to Python

apaszke commented on test/common.py:
```diff
@@ -152,6 +157,7 @@ def is_iterable(obj):
 class TestCase(unittest.TestCase):
     precision = 1e-5
     maxDiff = None
+    doCUDAMemoryCheck = True
```
Why is this set to `True` by default? Shouldn't it be only enabled in those test cases that actually deal with CUDA? Otherwise it's going to slow down a whole bunch of other things as well, and it might be entirely unnecessary. Our test suite is already unbearably slow, let's not make it 3x worse.

ssnl committed:
default doCUDAMemoryCheck to False

ezyang commented:
Maybe, to avoid having to fix all of the leaking kernels before we merge this, we should just blacklist all of the currently failing tests and then subsequently fix them.

ssnl committed:
make the wrapper skip-able

ssnl committed:
use TEST_MULTIGPU

ssnl committed:
add align_corners=True/False tests for Upsample; fix TEST_CUDNN

ssnl committed:
finalize interface

ssnl committed:
VolumetricMaxUnpooling_updateOutput

ssnl committed:
fix test_nccl

SsnL commented:
This PR also fixed the `test_cuda.py` running `TestTorch` problem, so the total test time increase on CUDA 9, CUDNN7, and py3 CI is 27s.

SsnL commented:
@apaszke @soumith Require a review! 

pr/caffe2-py2-gcc5-ubuntu16.04-test fail is unrelated and is being fixed.

ssnl committed:
rename THC caching allocator methods to be clearer

ssnl committed:
make the wrapped function a method

ezyang commented:
@pytorchbot retest this please

SsnL commented:
@pytorchbot retest this please

SsnL commented:
@pytorchbot retest this please

SsnL commented:
@pytorchbot retest this please

ezyang commented:
@pytorchbot retest this please

zou3519 commented on test/common_cuda.py:
```diff
@@ -0,0 +1,27 @@
+
+# Used below in `initialize_cuda_context_rng` to ensure that CUDA context and
+# RNG have been initialized.
+__cuda_ctx_rng_initialized = False
```
nit: does it make sense to hide this inside initialize_cuda_context_rng()? Something like
```
initialize_cuda_context_rng.__cuda_ctx_rng_initialized = False
```

SsnL replied:
I think that might be a bit confusing. We use this pattern elsewhere too (i.e. in `dataloader.py`).

SsnL commented:
@pytorchbot retest this please

ssnl committed:
address comments; revert changes to aten/src/THC/THCCachingAllocator.cpp

SsnL commented on test/common.py:
```diff
@@ -154,6 +177,41 @@ def is_iterable(obj):
     maxDiff = None
+    _do_cuda_memory_check = False
+
+    def __init__(self, method_name='runTest'):
```
Yes. A `TestCase` is initialized for each method.

zou3519 reviewed (approved):
lgtm, this is super cool!

ssnl committed:
fix renamed var
